libia2/include/ia2.h

```diff
@@ -142,11 +142,28 @@ asm(".macro mov_pkru_eax pkey\n"
 // compartments.
 #define IA2_SHARED_DATA __attribute__((section("ia2_shared_data")))
 
-#ifdef LIBIA2_INSECURE
-#define protect_tls_for_compartment(n)
-#else
-#define protect_tls_for_compartment(n) init_tls_##n();
-#endif
+// Initializes a compartment with protection key `n` when the ELF invoking this
+// macro is loaded. This must only be called once for each key. The compartment
+// includes all segments in the ELF except the `ia2_shared_data` section, if one
+// exists.
+#define _INIT_COMPARTMENT(n)                                                   \
+  extern int ia2_n_pkeys_to_alloc;                                             \
+  void ensure_pkeys_allocated(int *n_to_alloc);                                \
+  __attribute__((constructor)) static void init_pkey_##n##_ctor() {            \
+    ensure_pkeys_allocated(&ia2_n_pkeys_to_alloc);                             \
+    struct PhdrSearchArgs args = {                                             \
+        .pkey = n,                                                             \
+        .address = &init_pkey_##n##_ctor,                                      \
+    };                                                                         \
+    dl_iterate_phdr(protect_pages, &args);                                     \
+  }                                                                            \
+  void init_tls_##n(void) {                                                    \
+    struct PhdrSearchArgs args = {                                             \
+        .pkey = n,                                                             \
+        .address = &init_pkey_##n##_ctor,                                      \
+    };                                                                         \
+    dl_iterate_phdr(protect_tls_pages, &args);                                 \
+  }
```

Comment:
I missed that this got restored in a rebase, this is now in `ia2_compartment_init.inc`

Reply:
Good call, thanks for catching it.